Make more types jl_static_show readably

xal-0 opened the pull request:
Makes more types survive `jl_static_show`:
- Symbols
  - Symbols printed in the `:var"foo"` form use raw string escaping, fixing `:var"a\b"`, `:var"a\\"`, `:var"$a"`, etc.
  - Symbols that require parens use parens (`:(=)`, ...)
- Signed integers: Except for `Int`, signed integers print like `Int8(1)`.
- Floats: floats are printed in a naive but reversible (TODO: double check) way. `Inf(16|32|)` and `NaN(16|32|)` are printed, and `Float16`/`Float32` print the type (`Float32(1.5)`).  `Float64`s are printed with a trailing `.0` if it is necessary to disambiguate from `Int`.

Fixes #52677, https://github.com/JuliaLang/julia/issues/58484#issuecomment-2902468354, https://github.com/JuliaLang/julia/issues/58484#issuecomment-2898733637, and the specific case mentioned in #58484.  Improves the situation for #38902 but does not close it, because a few cases still do not round-trip (inexhaustive list):
- Non-canonical NaNs
- BFloat16
- User-defined primitive types.  This one is tricky, because they can have a size different from any type we have literals for.
- Without #40652 or something similar, precompile statements will still fail with field names.

xal-0 committed:
Make more types jl_static_show readably

xal-0 committed:
Typo fix

xal-0 committed:
Typo fix

KristofferC commented:
For https://github.com/JuliaLang/julia/pull/40652, can we thread something through to not show the field name in precompile printing contexts? Seems unnecessary to have the correctness of the printing be bike shredded on how people want it to print in the debugger.

xal-0 committed:
Fix jl_static_show for negative floats with no fractional part

vtjnash commented on src/rtutils.c:
```diff
@@ -777,6 +804,42 @@ static int jl_static_is_function_(jl_datatype_t *vt) JL_NOTSAFEPOINT {
+                              vt == jl_float32_type ? "32" :
+                                                      "";
+    // Digits required to print p significand bits: ceil(p * log(10, 2))
+    //   Float16   4
+    //   Float32   8
```
These seem unlikely correct, since the standard given results are one higher for these:
https://github.com/JuliaLang/julia/pull/57977/files#r2028868985
```suggestion
    //   Float16   5
    //   Float32   9
```

xal-0 replied:
Corrected formula from https://dl.acm.org/doi/pdf/10.1145/93542.93559: 

$N = 2 + \left\lfloor\frac{n}{\log_b B}\right\rfloor$

xal-0 replied:
```
function find_counterexample(T, prec, r)
    buf = zeros(UInt8, 32)
    for i=r
        x = reinterpret(T, i)
        isnan(x) && continue
        n = @ccall snprintf(buf::Ptr{UInt8}, 32::Csize_t, "%.*g"::Cstring; prec::Cint, x::Cdouble)::Csize_t
        ok, y = @ccall jl_try_substrtof(buf::Ptr{UInt8}, 0::Csize_t, n::Csize_t)::Tuple{Bool,Float32}
        ok && x === T(y) || return x
    end
end

@assert find_counterexample(Float16, 4, 0x0000:0xffff) === Float16(0.00010014)
@assert find_counterexample(Float16, 5, 0x0000:0xffff) === nothing

@assert find_counterexample(Float32, 8, 0x00000000:0xffffffff) === 1.00000075f-36
@assert find_counterexample(Float32, 9, 0x00000000:0xffffffff) === nothing
```

vtjnash commented on src/rtutils.c:
```diff
@@ -777,6 +804,42 @@ static int jl_static_is_function_(jl_datatype_t *vt) JL_NOTSAFEPOINT {
+        n += jl_printf(out, "%.17g%s", v, dotzero ? ".0" : "");
+    }
+    else if (vt == jl_float32_type) {
+        n += jl_printf(out, "Float32(%.8g)", v);
```
This can be implemented compactly similar to f64 case, with style '%e' and  `e` replaced with `f` in the buffer

xal-0 committed:
Correct jl_static_show_string raw escaping

n backslashes followed by " => 2n+1 backslashes and "
n backslahes at the end     => 2n backslashes

xal-0 committed:
Print floats with the correct precision, don't rely on fabs/modf

vtjnash commented on src/rtutils.c:
```diff
@@ -777,6 +812,55 @@ static int jl_static_is_function_(jl_datatype_t *vt) JL_NOTSAFEPOINT {
+        n += jl_printf(out, "%sInf%s", v < 0 ? "-" : "", size_suffix);
+    }
+    else if (vt == jl_float64_type) {
+        size_t m = snprintf(buf, sizeof buf, "%.17g", v);
+        jl_uv_puts(out, buf, m);
+        n += m;
+        // If there is no decimal point or exponent, disambiguate
+        if (!strpbrk(buf, ".e"))
+            jl_printf(out, ".0");
```
While clever, would it be better to use the standard flag `#` for this to preserve all trailing zeros (rather than to imply the trailing digit is `.0` which may not be entirely true given final digit rounding, though I think it would have to round-trip the same since 17 digits is full accuracy)
```suggestion
        n += jl_printf(buf, sizeof buf, "%#.17g", v);
```

vtjnash replied:
I assume that later printed as `Float16(-0)`, no? (though that makes it `isequal 0.0`, not `-0.0`)

xal-0 committed:
Update src/rtutils.c

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

xal-0 committed:
Update src/rtutils.c

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

xal-0 committed:
Fixes, additional tests, use trailing zeros for floats

xal-0 committed:
Fix -Werror=c++-compat

vtjnash commented on src/rtutils.c:
```diff
@@ -954,17 +1034,21 @@ static size_t jl_static_show_x_(JL_STREAM *out, jl_value_t *v, jl_datatype_t *vt
+        // Avoid unnecessary Int64(x)/Int32(x)
+        n += jl_printf(out, "%" PRIdPTR, *(intptr_t*)v);
+    }
     else if (vt == jl_int64_type) {
```
It isn't problematic, since that `Int128()` printing means a bitcast, not a function call

vtjnash replied:
Because it is human readable, and thus a useful debugging tool. Someone should implement a de-serializer if they want to use this for generating precompile statements.

kpamnany replied:
This is a disingenuous argument @vtjnash -- nobody reads this output for debugging (and we've tried); on the contrary, it is actively used for generating precompile statements. It would be better to remove all the "little gotchas".

KristofferC commented:
Backporting this to 1.11 gives:

```
Test Failed at /home/kc/julia1.11/test/show.jl:1562
  Expression: v === eval(Meta.parse(static_shown(v)))
   Evaluated: Float16(1.0) === Inf16

Test Failed at /home/kc/julia1.11/test/show.jl:1562
  Expression: v === eval(Meta.parse(static_shown(v)))
   Evaluated: Float16(1.5) === Inf16

Test Failed at /home/kc/julia1.11/test/show.jl:1562
  Expression: v === eval(Meta.parse(static_shown(v)))
   Evaluated: Float16(0.4893) === Inf16

Test Failed at /home/kc/julia1.11/test/show.jl:1562
  Expression: v === eval(Meta.parse(static_shown(v)))
   Evaluated: Float16(0.00010014) === Inf16
....
```

Am I missing some other PR that is needed?

KristofferC commented:
That should already be on 1.11 though?

topolarity commented:
Probably related to the Float16 intrinsic fixes that @xal-0 made in 1.12?
